### ISSUE TITLE
Various tests fail on Arch Linux

### ISSUE THREAD

**Issue by kpcyrd** [davidlattimore/wild#230]
hello, just wanted to let you know the following integration tests are currently failing for me:

I had to truncate some of the assembly output due to github's size limit - [full output.txt](https://github.com/user-attachments/files/18052332/output.txt)

```
     Running tests/integration_tests.rs (target/debug/deps/integration_tests-bccecb31354265dc)

running 31 tests
/usr/bin/ld: /home/user/repos/github/wild/wild/tests/build/weak-vars.16-187e5e57cedc9161.o: in function `_start':
weak-vars.c:(.text+0x171): undefined reference to `__stack_chk_fail'
test integration_test::program_name_06___weak_vars_c__ ... FAILED
test integration_test::program_name_05___data_c__ ... ok
test integration_test::program_name_01___trivial_c__ ... ok
test integration_test::program_name_04___global_definitions_c__ ... ok
test integration_test::program_name_10___init_test_c__ ... ok
test integration_test::program_name_12___internal_syms_c__ ... ok
test integration_test::program_name_03___link_args_c__ ... ok
test integration_test::program_name_14___old_init_c__ ... ok
test integration_test::program_name_11___ifunc_c__ ... ok
test integration_test::program_name_16___stack_alignment_s__ ... ok
test integration_test::program_name_02___trivial_main_c__ ... ok
test integration_test::program_name_15___custom_section_c__ ... ok
test integration_test::program_name_17___got_ref_to_local_c__ ... ok
test integration_test::program_name_18___local_symbol_refs_s__ ... ok
test integration_test::program_name_19___archive_activation_c__ ... ok
test integration_test::program_name_22___comments_c__ ... ok
test integration_test::program_name_21___string_merging_c__ ... ok
test integration_test::program_name_20___common_section_c__ ... ok
test integration_test::program_name_23___eh_frame_c__ ... ok
test integration_test::program_name_24___trivial_asm_s__ ... ok
test integration_test::program_name_25___non_alloc_s__ ... ok
error: failed to find a `codegen-backends` folder in the sysroot candidates:
       * /home/user/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu
       * /home/user/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu

test integration_test::program_name_27___rust_integration_rs__ ... FAILED
test integration_test::program_name_13___tls_c__ ... ok
test integration_test::program_name_31___input_does_not_exist_c__ ... ok
test integration_test::program_name_09___weak_fns_archive_c__ ... ok
test integration_test::program_name_07___weak_vars_archive_c__ ... ok
test integration_test::program_name_08___weak_fns_c__ ... ok
test integration_test::program_name_29___cpp_integration_cc__ ... ok
test integration_test::program_name_26___libc_integration_c__ ... FAILED
test integration_test::program_name_28___rust_integration_dynamic_rs__ ... ok
test integration_test::program_name_30___rust_tls_rs__ ... ok

failures:

---- integration_test::program_name_06___weak_vars_c__ stdout ----
Error: Failed to build program `weak-vars.c` with linker `ld` config `16`

Caused by:
    Linker failed. Relink with:
    /usr/bin/ld --gc-sections -static -z now -o /home/user/repos/github/wild/wild/tests/build/weak-vars.c-16.ld /home/user/repos/github/wild/wild/tests/build/weak-vars.16-187e5e57cedc9161.o /home/user/repos/github/wild/wild/tests/build/weak-vars1.16-404e2a27ec0a0590.o /home/user/repos/github/wild/wild/tests/build/exit.16-ec057211efac5d76.o

---- integration_test::program_name_27___rust_integration_rs__ stdout ----
Error: Failed to build program `rust-integration.rs` with linker `ld` config `cranelift-static`

Caused by:
    Compilation failed: rustc +nightly -C linker=clang -C link-arg=--ld-path=/home/user/repos/github/wild/target/debug/wild /home/user/repos/github/wild/wild/tests/sources/rust-integration.rs -Zcodegen-backend=cranelift --target x86_64-unknown-linux-musl -C relocation-model=static -C target-feature=+crt-static -C debuginfo=2 --cfg cranelift

---- integration_test::program_name_26___libc_integration_c__ stdout ----
wild: /home/user/repos/github/wild/wild/tests/build/libc-integration.c-gcc-static-pie.wild
ld: /home/user/repos/github/wild/wild/tests/build/libc-integration.c-gcc-static-pie.ld
asm._IO_file_seek
  ORIG            No layout information in range 84690..8469d (has Some(b28..d83f7))
                  endbr64
                  mov 0x70(%rdi),%edi
  
  wild 0x00084697 ff 25 1b 7b 04 00 jmpq *0x47B28  // Jmp_rm64(0x300cc1b8) GOT(__lseek64) GOT(lseek64) GOT(__libc_lseek64) GOT(__libc_lseek) GOT(__lseek) GOT(lseek)
  ld   0x00031af7 e9 c4 34 01 00 jmp 0x00000000000134D0  // Jmp_rel32_64(0x30044fc0) __lseek __lseek64 lseek __libc_lseek64 __libc_lseek lseek64
  ld   0x00031afc 90 nop  // Nopd()
  ORIG            ff 25 00 00 00 00 jmpq *0xD  // R_X86_64_GOTPCRELX -> `__lseek64` -4
  TRACE           value_flags=ADDRESS | CAN_BYPASS_GOT resolution_flags=DIRECT | GOT
  

asm.__deregister_frame
  ORIG            No layout information in range 3a7d0..3a7f3 (has Some(b28..d83f7))
                  endbr64
                  mov (%rdi),%eax
                  test %eax,%eax
                  jne 0x0000000000000010
                  ret
                  nopl (%rax,%rax)
                  push %rbp
                  mov %rsp,%rbp
                  call 0xFFFFFFFFFFFFF430
                  pop %rbp
                  mov %rax,%rdi
  
  wild 0x0003a7ed ff 25 45 18 09 00 jmpq *0x91868  // Jmp_rm64(0x300cc038) GOT(__libc_free) GOT(free) GOT(__free)
  ld   0x0009932d e9 1e c1 f7 ff jmp 0xFFFFFFFFFFF7C140  // Jmp_rel32_64(0x30015450) __libc_free __free free
  ld   0x00099332 90 nop  // Nopd()
  ORIG            ff 25 00 00 00 00 jmpq *0x23  // R_X86_64_GOTPCRELX -> `free` -4
  TRACE           value_flags=ADDRESS | CAN_BYPASS_GOT resolution_flags=DIRECT | GOT
  

asm.__dl_libc_freemem
  ORIG            No layout information in range 70010..70139 (has Some(b28..d83f7))
                  endbr64
                  push %rbp
                  mov %rsp,%rbp
                  push %r12
                  push %rbx
                  mov 0,%rbx  // 0xAAA=_dl_all_dirs
                  cmp 0,%rbx  // 0xAAA=_dl_init_all_dirs
                  je 0x0000000000000035
                  nopl (%rax,%rax)
                  mov %rbx,%rdi
                  mov (%rbx),%rbx
                  call 0  // 0xAAA=free
                  cmp %rbx,0  // 0xAAA=_dl_init_all_dirs
                  jne 0x0000000000000020
                  cmpq $0,0  // 0xAAA=_dl_nns
                  je 0x00000000000000B9
                  mov 0,%r12  // 0xAAA=_dl_ns
                  test %r12,%r12
                  je 0x00000000000000AF
                  nopl (%rax,%rax)
                  mov 0x38(%r12),%rax


[ TRUNCATED ]


                  call 0  // 0xAAA=pthread_mutex_lock
                  lea 0,%rdi  // 0xAAA=version_lock_cond
                  call 0  // 0xAAA=pthread_cond_broadcast
                  mov %rbx,%rdi
                  mov -8(%rbp),%rbx
                  leave
  
  wild 0x000385a6 ff 25 84 3a 09 00 jmpq *0x93AD0  // Jmp_rm64(0x300cc030) GOT(___pthread_mutex_unlock) GOT(pthread_mutex_unlock) GOT(__pthread_mutex_unlock)
  ld   0x000970e6 e9 d5 8c f7 ff jmp 0xFFFFFFFFFFF78D20  // Jmp_rel32_64(0x3000fdc0) __pthread_mutex_unlock pthread_mutex_unlock ___pthread_mutex_unlock
  ld   0x000970eb 90 nop  // Nopd()
  ORIG            ff 25 00 00 00 00 jmpq *0x4C  // R_X86_64_GOTPCRELX -> `pthread_mutex_unlock` -4
  TRACE           value_flags=ADDRESS | CAN_BYPASS_GOT resolution_flags=DIRECT | GOT
  


Error: Validation failed.
Binary `/home/user/repos/github/wild/wild/tests/build/libc-integration.c-gcc-static-pie.wild`. Relink with:
WILD_WRITE_LAYOUT=1 WILD_WRITE_TRACE=1 OUT=/home/user/repos/github/wild/wild/tests/build/libc-integration.c-gcc-static-pie.wild /home/user/repos/github/wild/wild/tests/build/libc-integration.c-gcc-static-pie.save/run-with cargo run --bin wild --
 To revalidate:
cargo run --bin linker-diff -- --wild-defaults --ignore '.got.plt,.dynamic.DT_PLTGOT,.dynamic.DT_JMPREL,.dynamic.DT_NEEDED,.dynamic.DT_PLTREL,section.plt.entsize,section.rodata.cst32.entsize,section.rela.plt.link' --ref /home/user/repos/github/wild/wild/tests/build/libc-integration.c-gcc-static-pie.ld /home/user/repos/github/wild/wild/tests/build/libc-integration.c-gcc-static-pie.wild


failures:
    integration_test::program_name_06___weak_vars_c__
    integration_test::program_name_26___libc_integration_c__
    integration_test::program_name_27___rust_integration_rs__

test result: FAILED. 28 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.35s

error: test failed, to rerun pass `-p wild --test integration_tests`
```

(fyi I also noticed there's more failures when testing with a system-wide installed Rust install):

<details>

```
error: multiple input filenames provided (first two filenames are `+nightly` and `/build/wild/src/wild-0.3.0/wild/tests/sources/rust-integration.rs`)
error: multiple input filenames provided (first two filenames are `+nightly` and `/build/wild/src/wild-0.3.0/wild/tests/sources/rust-tls.rs`)


error: multiple input filenames provided (first two filenames are `+nightly` and `/build/wild/src/wild-0.3.0/wild/tests/sources/rust-integration-dynamic.rs`)
```

</details>

thanks!

**Comment by davidlattimore** [comment]
Thanks for the report. I suspect there's at least two different things happening.

```
  wild 0x00084697 ff 25 1b 7b 04 00 jmpq *0x47B28  // Jmp_rm64(0x300cc1b8) GOT(__lseek64) GOT(lseek64) GOT(__libc_lseek64) GOT(__libc_lseek) GOT(__lseek) GOT(lseek)
  ld   0x00031af7 e9 c4 34 01 00 jmp 0x00000000000134D0  // Jmp_rel32_64(0x30044fc0) __lseek __lseek64 lseek __libc_lseek64 __libc_lseek lseek64
  ld   0x00031afc 90 nop  // Nopd()
  ORIG            ff 25 00 00 00 00 jmpq *0xD  // R_X86_64_GOTPCRELX -> `__lseek64` -4
  TRACE           value_flags=ADDRESS | CAN_BYPASS_GOT resolution_flags=DIRECT | GOT
```

This failure looks like GNU ld is performing an optimisation that we don't support yet. What version of `ld` and `gcc` do you have? Also, Linux distribution are you running?

```
---- integration_test::program_name_27___rust_integration_rs__ stdout ----
Error: Failed to build program `rust-integration.rs` with linker `ld` config `cranelift-static`
```

This failure is because you don't have the rustc-cranelift backend installed. There's instructions for installing it in `CONTRIBUTING.md` - although if you're using your distro's rustc rather than rustup, I'm not sure how practical it'd be.

**Comment by kpcyrd** [comment]
I'm building this on Arch Linux with the following versions:

```
% ld --version
GNU ld (GNU Binutils) 2.43.0
Copyright (C) 2024 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```

```
% gcc --version
gcc (GCC) 14.2.1 20240910
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
% lld --version
lld is a generic driver.
Invoke ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld (WebAssembly) instead
```

```
% clang --version
clang version 18.1.8
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

I've briefly looked into [rustc-cranelift](https://github.com/rust-lang/rustc_codegen_cranelift) but since there's no release tags beyond v0.1.0 from 2020 I'd disable that test instead of trying to package it.

**Comment by mati865** [comment]
I have the same failures on Arch Linux (06 and 26), but adding `-fstack-protector` solves the problem with `program_name_06___weak_vars_c__`.
Proposed the fix: https://github.com/davidlattimore/wild/pull/281

**Comment by davidlattimore** [comment]
I added a dockerfile to make it easier for those of us that don't run Alpine natively to have a look. In that image, I see the following tests fail:

* integration_test::program_name_02___trivial_main_c__
  * `Insufficient .gnu.version allocation` - For some reason we're attempting to write more data to .gnu.version than we allocated during the layout phase.
* integration_test::program_name_11___ifunc_c__
  * Ifuncs aren't supported on musl. I guess we need some way to skip this test for musl.
* integration_test::program_name_32___ifunc2_c__
  * Same as above
* integration_test::program_name_26___libc_integration_c__
  * Ifuncs again. Probably don't want to skip the whole test, but just the part of this test that's related to ifuncs.
* integration_test::program_name_27___rust_integration_rs__
  * Initially failed because I forgot to add `clang` to the installs in the dockerfile.
  * After that, this is failing because we're trying to run `bash`. Need to either switch to `sh` or maybe install bash in the image.

The following 3 tests also failed for me, but I didn't look at them, since I figured they'd be more of the same.

* integration_test::program_name_28___rust_integration_dynamic_rs__
* integration_test::program_name_29___cpp_integration_cc__
* integration_test::program_name_30___rust_tls_rs__

I probably won't get a chance to look more at any of this for a while, since I'm focusing my time on the linker-diff rewrite. If anyone would like to have a go at fixing any of these and needs more guidance, I'm definitely happy to help where I can.

**Comment by mati865** [comment]
Are you sure you meant to write in this issue? All previous comments were regarding issues on Arch Linux, and issues on Alpine Linux don't seem to be related.

**Comment by davidlattimore** [comment]
Oops, you're right. Nobody was talking about Alpine until I went and got my distributions-that-start-with-a mixed up. I'll open a separate issue for Alpine.

**Comment by mati865** [comment]
There are basically two failures left here, both related to `program_name_06___weak_vars_c__` test:
- missing optimisation of indirect jump to relative (I have WIP change for this one; passes wild's tests but needs a bit of refinement)
- outputting unnecessary `.plt.got` section:
```
---- integration_test::program_name_27___libc_integration_c__ stdout ----
wild: /home/mateusz/projects/wild/wild/tests/build/libc-integration-0.clang-initial-exec-59c69d78d0f4ce31.wild.so
ld: /home/mateusz/projects/wild/wild/tests/build/libc-integration-0.clang-initial-exec-59c69d78d0f4ce31.ld.so
lld: /home/mateusz/projects/wild/wild/tests/build/libc-integration-0.clang-initial-exec-59c69d78d0f4ce31.lld.so
section.plt.got
  wild OK
  ld Section missing
  lld Section missing

section.gnu_debuglink
  wild Section missing
  ld Section missing
  lld OK


Error: Validation failed.
```

**Comment by mati865** [comment]
FYI my WIP branch that passes all tests on Arch is here: https://github.com/mati865/wild/tree/push-wlnuolzyswxz

**Comment by davidlattimore** [comment]
Thanks for working on that. Those changes look reasonable to me. If you'd like to create a PR, I'd be happy to merge it.

**Comment by marxin** [comment]
> FYI my WIP branch that passes all tests on Arch is here: https://github.com/mati865/wild/tree/push-wlnuolzyswxz

That's great! Note that you can easily add a CI tester for the docker image, and it will be tested in the future.

**Comment by mati865** [comment]
I started digging into my todos, and things look pretty wild there (pun intended).
`.gnu_debuglink` should be easily solvable, and it's totally fine to ignore it for now.
`.plt.got` on the other hand... I plan to prepare a short report what I had found so far once I'm back to my PC and find some time.

**Comment by mati865** [comment]
TL;DR
I'll avoid `.gnu_debuglink` error by skipping it in linker-diff, but I have no idea how to deal with the second one.

---

First, keep in mind this is the work of a newbie in the area of ELF internals. I could have made big mistakes, so take it with a grain of salt.

I'll split this into two parts based on the error: `.gnu_debuglink` and `.plt.got`.
I have navigated to `tests/build/libc-integration.c-clang-initial-exec.save` and prepared `<linker>-bin` files using `run-with`.

Let's start with the easy part:
```
❯ objdump -Wk -WN wild-bin

wild-bin:     file format elf64-x86-64

❯ objdump -Wk -WN lld-bin

lld-bin:     file format elf64-x86-64

Contents of the .gnu_debuglink section:

  Separate debug info file: Scrt1.o.debug
  CRC value: 0x126e8809
objdump: Warning: There are 0x54 extraneous bytes at the end of the section

❯ objdump -Wk -WN Scrt1.o

Scrt1.o:     file format elf64-x86-64

Contents of the .gnu_debuglink section:

  Separate debug info file: Scrt1.o.debug
  CRC value: 0x126e8809
```
What can we tell from that?
Arch Linux has integrated support for [Debuginfod](https://wiki.archlinux.org/title/Debuginfod), so debuggers can download necessary split debug info on the go. Cool!
I think all the linker has to do is to here is to copy the section from the objects, as the compiler is the one responsible for splitting debug info.

Definitely nice to have but IMO not very important at this stage of development. Could be ignored in the linker-diff to avoid the failure and be extracted into a new issue. Perhaps with the "good first issue" label? I'm not sure if there is something more to do here.

---

Now let's look at `.plt.got` that is present only with wild but absent with LLD and ld.bfd:
```
❯ readelf -S wild-bin
  [Nr] Name              Type             Address           Offset
[..]
  [16] .plt.got          PROGBITS         0000000000002630  00001630
       0000000000000110  0000000000000010  AX       0     0     16
[..]
  [20] .got              PROGBITS         0000000000003bc8  00001bc8
       00000000000000e0  0000000000000008  WA       0     0     8
[..]

❯ readelf -S lld-bin
[..]
  [10] .rela.plt         RELA             0000000000000ac0  00000ac0
       0000000000000198  0000000000000018  AI       4    24     8
[..]
  [17] .plt              PROGBITS         00000000000021b0  000011b0
       0000000000000120  0000000000000000  AX       0     0     16
[..]
  [23] .got              PROGBITS         00000000000034c8  000014c8
       0000000000000048  0000000000000000  WA       0     0     8
  [24] .got.plt          PROGBITS         0000000000003510  00001510
       00000000000000a0  0000000000000000  WA       0     0     8
[..]

❯ readelf -S bfd-bin
[..]
  [11] .rela.plt         RELA             0000000000000b80  00000b80
       0000000000000198  0000000000000018  AI       6    24     8
[..]
  [13] .plt              PROGBITS         0000000000001020  00001020
       0000000000000120  0000000000000010  AX       0     0     16
[..]
  [24] .got              PROGBITS         0000000000003f10  00002f10
       00000000000000f0  0000000000000008  WA       0     0     8
[..]
```
Okay, relocations again...
Let's compare them:
| value | type wild | type LLD | type ld.bfd |
| -------------------|--------------------|------------------|------------------ |
| get_tvar_local@Base | R_X86_64_GLOB_DAT | R_X86_64_64 | R_X86_64_64 |
| get_tvar_local2@Base | R_X86_64_GLOB_DAT | R_X86_64_64 | R_X86_64_64 |
| get_weak_var@Base | R_X86_64_GLOB_DAT | R_X86_64_64 | R_X86_64_64 |
| get_weak_var2@Base | R_X86_64_GLOB_DAT | R_X86_64_64 | R_X86_64_64 |
| weak_fn1@Base | R_X86_64_GLOB_DAT | R_X86_64_GLOB_DAT | R_X86_64_GLOB_DAT |
| weak_fn1@Base | R_X86_64_GLOB_DAT | R_X86_64_64 | R_X86_64_64 |
| weak_fn2@Base | R_X86_64_GLOB_DAT | R_X86_64_GLOB_DAT | R_X86_64_GLOB_DAT |
| get_weak_var@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| tvar3@Base | R_X86_64_TPOFF64 | R_X86_64_TPOFF64 | R_X86_64_TPOFF64 |
| get_weak_var2@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| set_tvar_local@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| get_weak_tvar@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| value42@Base | R_X86_64_GLOB_DAT | R_X86_64_GLOB_DAT | R_X86_64_GLOB_DAT |
| black_box@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| get_weak_tvar2@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| get_tvar_local@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| get_tvar_local2@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| compute_value10@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| set_tvar2@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| get_42@Base | R_X86_64_GLOB_DAT | R_X86_64_JUMP_SLOT | R_X86_64_JUMP_SLOT |
| tvar2@@Base | - | - | R_X86_64_TPOFF64 |

I'm still reading and trying to understand the docs linked in this repo, but from what I already learned, wild seems to do well in this case: https://maskray.me/blog/2021-08-29-all-about-global-offset-table#combining-.got-and-.got.plt
Or maybe `-ftls-model=initial-exec` somehow makes the difference here?
Clearly there is some lack of knowledge on my part here, but maybe symbol tables will tell more?

I'm not going to post them here, but they are vastly different.

The main takeaways here are:
- in `.dynsym` table, wild unlike LLD and ld.bfd preserved sizes for the undefined functions. I can see no other major differences. Sounds fine, I guess?
- in `.symtab` there is so much things going on, I don't even know what to look at (nor do I understand it).  Just for perspective, wild created 16 entries, LLD 44 entries and ld.bfd 49 entries.

Analysis of the assembly is beyond my skill level, so I'll leave that for somebody who wants to go deeper (or maybe even a future me?). Attaching the outputs from my system and also tools versions:
```
❯ ld.lld --version
LLD 19.1.7 (compatible with GNU linkers)

tests/build/libc-integration.c-clang-initial-exec.save on  HEAD (492caee)
❯ ld.bfd --version
GNU ld (GNU Binutils) 2.43.1
Copyright (C) 2024 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
```




[clang-initial-exec.tar.gz](https://github.com/user-attachments/files/18550799/clang-initial-exec.tar.gz)

**Comment by davidlattimore** [comment]
Thanks for the thorough investigation!

GNU ld and lld have various differences in how they name the different PLT sections, what stuff they put in which section etc. To an extent, these differences are cosmetic and don't really matter, so just suppressing them is probably the right thing to do. At some point we should probably revisit how linker-diff diffs these kinds of section headers.

**Comment by davidlattimore** [comment]
The symbol differences are interesting. I thought all our tests passed -znow to the linker, but maybe this test is overriding that, since I wouldn't expect to see jump slots used unless lazy binding was active (wild doesn't support -zlazy, but the other linkers default to it).

As for the glob-dat vs x86_64 relocations, I don't think I'd seen x86_64 used as a runtime relocation, only in the input object files, so that's interesting. I think they're equivalent, but maybe there's some difference that I'm unaware of.

**Comment by mati865** [comment]
With #356 and updated linker-diff there seems to be only one last failure:
```
rel.R_X86_64_PC32.R_X86_64_GOTPCRELX
  `/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/14.2.1/../../../../lib64/libc.a` @ `pthread_create.o` .text start_thread
  ORIG 0x00469: [ ff 15 00 00 00 00 ] callq *0x46F
                        ^^^^^^^^^^^ R_X86_64_GOTPCRELX
                                    __call_tls_dtors -4
  wild 0x88389: [ 67 e8 71 7c f7 ff ] call 0
                        ^^^^^^^^^^^ R_X86_64_PC32 CallIndirectToRelative
                                    0x0
  wild TRACE: message=relaxation applied kind=CallIndirectToRelative
              value_flags=ABSOLUTE | CAN_BYPASS_GOT resolution_flags=DIRECT | GOT
              rel_info.kind=Relative symbol_name=__call_tls_dtors
  ld   0x0d829: [ ff 15 c1 f7 0b 00 ] callq *0xCCFF0
                        ^^^^^^^^^^^ R_X86_64_GOTPCRELX NoOp
                                    GOT->#0x0
  lld  0x37fe9: [ ff 15 71 47 09 00 ] callq *0xCC760
                        ^^^^^^^^^^^ R_X86_64_GOTPCRELX NoOp
                                    GOT->#0x0
```
Wild applied relaxation where other linkers didn't (maybe I should diff with mold as well?). I cannot tell if this is a valid or expected optimisation.
This happens for `{clang,gcc}-libc-integration.c-clang-static-pie` tests.

**Comment by davidlattimore** [comment]
Ah, interesting. Calling 0 directly and calling 0 indirectly via the GOT are both undefined behaviour. In that sense, they're definitely equivalent. I guess there's two options (1) make Wild not optimise this case and do the same thing as the other linkers or (2) make linker-diff accept these two things as equivalent.

If you wanted to experiment with removing the optimisation, the code to look at is `value_flags_from_elf_symbol` where it sets the flag `CAN_BYPASS_GOT`. For dynamically linked executables, we bypass the GOT if the symbol is going to be defined in the executable, since there's no way for a shared library to override a symbol provided by the executable. For undefined symbols in a dynamically linked executable, we don't bypass the GOT, since it's possible that a dynamic library might provide the symbol at runtime. For statically linked executables, i.e. the case that's failing here, we always allow GOT bypass regardless of whether the symbol is defined or not, since it's statically linked, so there won't be any shared libraries and the symbol can never be defined at runtime. It's possible that this rule regarding statically linked executables could be removed - i.e. something like the following:

```patch
diff --git a/libwild/src/symbol_db.rs b/libwild/src/symbol_db.rs
index e47e529686..2f16c2aa2f 100644
--- a/libwild/src/symbol_db.rs
+++ b/libwild/src/symbol_db.rs
@@ -499,7 +499,6 @@
     let is_undefined = sym.is_undefined(LittleEndian);
     let mut can_bypass_got = sym.st_visibility() != object::elf::STV_DEFAULT
         || sym.is_local()
-        || args.output_kind().is_static_executable()
         // Symbols defined in an executable cannot be interposed since the executable is always the
         // first place checked for a symbol by the dynamic loader.
         || (args.output_kind().is_executable() && !is_undefined);
@@ -513,7 +512,7 @@
     } else if sym.st_type() == object::elf::STT_GNU_IFUNC {
         ValueFlags::IFUNC
     } else if is_undefined {
-        if can_bypass_got {
+        if can_bypass_got || args.output_kind().is_static_executable() {
             ValueFlags::ABSOLUTE
         } else {
             // If we can't bypass the GOT, then an undefined symbol might be able to be defined at
```

The second line of the change is so that we don't attempt to create a dynamic symbol for the undefined symbol as that wouldn't work in a static executable. However, this change breaks other things. It seems that when there's a `PLT32` relocation to an undefined symbol, the other linkers bypass the PLT. Wild doesn't currently have a separate concept by being able to bypass the PLT but not bypass the GOT. It might be possible to achieve this effect by adjusting the code that selects relaxations as follows:

```patch
diff --git a/libwild/src/x86_64.rs b/libwild/src/x86_64.rs
index 85e94dfb0e..a9fa8acb0d 100644
--- a/libwild/src/x86_64.rs
+++ b/libwild/src/x86_64.rs
@@ -198,7 +198,7 @@
                     }
                     _ => {}
                 }
-                if can_bypass_got {
+                if can_bypass_got && !(is_absolute && output_kind.is_static_executable()) {
                     match section_bytes.get(offset - 2..offset)? {
                         // call *x(%rip)
                         [0xff, 0x15] => {
```

Here, we're using the `ABSOLUTE` flag to determine that the symbol is likely undefined - either that or we're calling a symbol with an absolute value, which would be weird. This at least doesn't appear to break any tests on my machine. I haven't looked further to see if this does what's desired though.

There's not yet any cases in linker-diff where we make certain optimisations optional, so it's a bit of an underexplored space as yet. However, I think we probably do want something along those lines. I'd suggesting making the rule as tight as possible. i.e. allow `R_X86_64_PC32, CallIndirectToRelative, 0x0` to match `R_X86_64_GOTPCRELX NoOp, GOT->#0x0`. This can be done in `Resolution::matches`. We could optionally make the rule only apply if the binary being diffed is a static executable, although that would require that information to be passed into the `matches` method. I don't think it's necessary to check the file type, since the fact that one side of the match resolved to `GOT->#0x0` means that there's no dynamic relocation. In the case of a shared object or a dynamically linked executable, I'd expect there to be a dynamic relocation, so the rule wouldn't match.

**Comment by mati865** [comment]
> Ah, interesting. Calling 0 directly and calling 0 indirectly via the GOT are both undefined behaviour. In that sense, they're definitely equivalent.

Oh, that's nice. Well, having UB itself isn't nice, but it opens up some possibilities.

I think it would be nice if linker-diff could detect the UBs and ignore what happens to that instruction (kinda do what compilers often do with UB). Or at least ignore cases when wild relaxes them more than other linkers.

It should be doable for executables with local symbols, but other cases would more or less be tricky. So it might be not terribly useful in general. Also, I have no idea how many UBs there are in the linker world.

> Here, we're using the `ABSOLUTE` flag to determine that the symbol is likely undefined - either that or we're calling a symbol with an absolute value, which would be weird. This at least doesn't appear to break any tests on my machine. I haven't looked further to see if this does what's desired though.

On Arch Linux this causes `program_name_08___weak_fns_c__` and `program_name_09___weak_fns_archive_c__` with the opposite error than previously 🙃.
This time ld.bfd optimised the UB but wild did not:
```
rel.R_X86_64_PLT32.R_X86_64_PC32
  `/home/mateusz/projects/wild/wild/tests/build/weak-fns-archive.12-a259cc1488953534.o` .text _start
  ORIG 0x000067: [ e8 00 00 00 00 ] call 0x000000000000006C
                      ^^^^^^^^^^^ R_X86_64_PLT32
                                  weak_fn5 -4
  wild 0x401697: [ e8 84 ff ff ff ] call 0x0000000000401620
                      ^^^^^^^^^^^ R_X86_64_PLT32 NoOp
                                  PLT->GOT->0x0
  wild TRACE: message=relocation applied value_flags=ABSOLUTE
              resolution_flags=DIRECT | GOT | PLT rel_info.kind=PltRelative
              symbol_name=weak_fn5
  ld   0x401067: [ e8 94 ef bf ff ] call 0
                      ^^^^^^^^^^^ R_X86_64_PC32 NoOp
                                  0x0
```
Surely, with enough conditions wild would reach fragile parity with ld.bfd, but I'm not convinced by such an idea.

Option 2 sounds good to me, but I'm not going to explore it in the immediate future because I still got few TODOs in this and other projects that I want to tackle first. However, if somebody wants, feel free to fix it one way or the other.

**Comment by davidlattimore** [comment]
Prior to the recent rewrite, linker-diff did have some code for detecting certain classes of undefined behaviour. If one of the reference linkers produced code that was UB, then it would allow that to match against anything that Wild produced. However, for the most part, that requires decoding the instructions and the new version of linker diff doesn't decode during matching, only when reporting a diff. I'd like if possible to avoid decoding instructions during diffing.

I changed linker-diff to accept this optimisation. I ended up not doing it in the `matches` function after all. When I attempted to do so, I realised that the things I needed to match on were arch-specific, so the rule itself would have been arch specific. Instead I ended up changing the way diff keys are computed for missing / extra optimisations, so now we can have a wild-specific diff-ignore that says that that extra optimisation is OK. I'm happier with this approach than what I originally proposed.

**Comment by mati865** [comment]
Thank you. The tests pass now!
I'll take a look into the PR in the morning.
